pkg/controller/populators/forklift-populator.go

```diff
@@ -0,0 +1,593 @@
+		var pvcs corev1.PersistentVolumeClaimList
+		matchingFields := client.MatchingFields{
+			dataSourceRefField: getPopulatorIndexKey(apiGroup, sourceKind, obj.GetNamespace(), obj.GetName()),
+		}
+		if err := mgr.GetClient().List(context.TODO(), &pvcs, matchingFields); err != nil {
+			log.Error(err, "Unable to list PVCs", "matchingFields", matchingFields)
+			return reqs
+		}
+		for _, pvc := range pvcs.Items {
+			reqs = append(reqs, reconcile.Request{NamespacedName: types.NamespacedName{Namespace: pvc.Namespace, Name: pvc.Name}})
+		}
+		return reqs
+	}
+
+	if err := c.Watch(source.Kind(mgr.GetCache(), sourceType),
+		handler.EnqueueRequestsFromMapFunc(mapDataSourceRefToPVC),
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reconcile the reconcile loop for the PVC with DataSourceRef of OvirtVolumePopulator or OpenstackVolumePopulator kind
+func (r *ForkliftPopulatorReconciler) Reconcile(_ context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("PVC", req.NamespacedName)
+	log.V(1).Info("reconciling Forklift PVCs")
+	return r.reconcile(req, r, log)
+}
+
+func (r *ForkliftPopulatorReconciler) reconcile(req reconcile.Request, populator populatorController, log logr.Logger) (reconcile.Result, error) {
+	pvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(context.TODO(), req.NamespacedName, pvc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if pvc.Status.Phase == corev1.ClaimBound {
+		r.log.Info("PVC is bound, skipping...", "pvc", pvc.Name)
+		return reconcile.Result{}, nil
+	}
+
+	// We first perform the common reconcile steps.
+	// We should only continue if we get a valid PVC'
+	pvcPrime, err := r.reconcileCommon(pvc, populator, log)
+	if err != nil || pvcPrime == nil {
+		return reconcile.Result{}, err
+	}
+
+	r.log.V(1).Info("reconciling PVC prime", "pvc", pvcPrime.Name)
+
+	// Each populator reconciles the target PVC in a different way
+	if cc.IsUnbound(pvc) || !cc.IsPVCComplete(pvc) {
+		return populator.reconcileTargetPVC(pvc, pvcPrime)
+	}
+
+	return r.reconcileCleanup(pvcPrime)
+}
+
+// TODO(benny) rename
+func (r *ForkliftPopulatorReconciler) reconcileCommon(pvc *corev1.PersistentVolumeClaim, populator populatorController, log logr.Logger) (*corev1.PersistentVolumeClaim, error) {
+	if pvc.DeletionTimestamp != nil {
+		log.V(1).Info("PVC being terminated, ignoring")
+		return nil, nil
+	}
+
+	log.V(1).Info("reconciling PVC prime")
+	pvcPrime, err := r.getPVCPrime(pvc)
+	if err != nil {
+		return nil, err
+	}
+	if pvcPrime != nil {
+		return pvcPrime, nil
+	}
+
+	dataSourceRef := pvc.Spec.DataSourceRef
+
+	if !isPVCForkliftKind(pvc) {
+		log.V(1).Info("reconciled unexpected PVC, ignoring")
+		return nil, nil
+	}
+
+	// TODO: Remove this check once we support cross-namespace dataSourceRef
+	if dataSourceRef.Namespace != nil {
```

Comment:
What about if the namespace is the same namespace as the PVC, that would not be cross namespace right?

Reply:
I took this from https://github.com/kubevirt/containerized-data-importer/blob/55dd0d0110754e17af3473a2c5622dcc10c90b2f/pkg/controller/populators/populator-base.go#L328

Reply:
I think that is wrong as well, it should check if != nil and not equals the namespace of the PVC.

Reply:
Added, @alromeros can you confirm?